### PR TITLE
Add per-rule disable configuration through json file

### LINKED ISSUE #20: Add per-rule configurations
## Problem

The plugin runs every rule in `allRules` with no way for a consumer to disable a subset. Real projects may want, for example, `vague-alt-text`
and `placeholder-alt-text` but not `repeated-alt` (which can be noisy on icon-heavy pages — see #19).

There's also no per-image escape hatch for false positives.

## Proposed scope

Two related but separable capabilities:

1. **Per-rule enable/disable.** Some way for the consumer to pass an allow-list or deny-list of rule IDs to the plugin. Mechanisms to consider:
   - Plugin args via the scanner action (e.g., a YAML input)
   - Environment variable (e.g., `ALT_TEXT_SCAN_DISABLED_RULES=repeated-alt`)
   - Config file in `.github/scanner-plugins/alt-text-scan/`
2. **Per-image opt-out.** A documented `data-*` attribute or `aria-label` convention the author can add to an `` to suppress findings on it (across all rules, or for a specific rule ID).

## Acceptance criteria

- Document the chosen mechanism in the README (`Configuration` section).
- Add tests proving disabled rules emit zero findings.
- Update `PLUGINS.md` in the scanner repo if the mechanism is shared.

### PR DESCRIPTION
**Pull request by kzhou314**
Closes #20.

Adds opt-out configuration for individual rules. Consumers create `.github/scanner-plugins/alt-text-scan/config.json` in their scanner repo:

```json
{
  "rules": {
    "repeated-alt-text": false,
    "placeholder-alt-text": false
  }
}

### CODE REVIEW

**Comment by JoyceZhu** [comment]
Something to consider here: if we end up with different defaults for different rules (in other words, some rules are enabled by default while others are disabled by default), you'd need an `enabledRules` as well, and the set logic would then not be as obvious to users just by reading the config file.

I'd consider some combination of the following:

- Have a file named `default-config.json` or even `default.json` which explicitly spells out which rules are on by default
- Adjust the JSON schema for overrides to be `"rulename": true/false`. This is how _many_ if not almost all boolean settings are described in the [JSON file version of VSCode's settings customization](https://code.visualstudio.com/docs/configure/settings#_settings-json-file)
- If we potentially want to allow other customization beyond allowing fine-grained rule adjustment, you could even have the schema be something like
```
{
  'rules': {
     'vague-alt-text': true,
     'placeholder-alt-text': false
  }
}
``` 
to further separate it out.

**Comment by kzhou314** [comment]
> Something to consider here: if we end up with different defaults for different rules (in other words, some rules are enabled by default while others are disabled by default), you'd need an `enabledRules` as well, and the set logic would then not be as obvious to users just by reading the config file.
> 
> I'd consider some combination of the following:
> 
> * Have a file named `default-config.json` or even `default.json` which explicitly spells out which rules are on by default
> * Adjust the JSON schema for overrides to be `"rulename": true/false`. This is how _many_ if not almost all boolean settings are described in the [JSON file version of VSCode's settings customization](https://code.visualstudio.com/docs/configure/settings#_settings-json-file)
> * If we potentially want to allow other customization beyond allowing fine-grained rule adjustment, you could even have the schema be something like
> 
> ```
> {
>   'rules': {
>      'vague-alt-text': true,
>      'placeholder-alt-text': false
>   }
> }
> ```
> 
> to further separate it out.

That's a good observation to make! Definitely better accommodates the stretch goals later on. I've adjusted the schema in the most recent commit to use your idea.

File: src/config.ts [hunk 1]
```diff
@@ -10,7 +10,10 @@ export async function loadConfig(configPath: string, knownRuleIds: ReadonlySet<s
   let raw: string
   try {
     raw = await readFile(configPath, 'utf8')
-  } catch {
+  } catch (err) {
+    if ((err as NodeJS.ErrnoException)?.code !== 'ENOENT') {
```
Comment (JoyceZhu) [inline src/config.ts]: technically, this caught error could be _anything_ (TypeScript probably told you it was type `unknown`), so I'd recommend [a bit more code here](https://medium.com/with-orus/the-5-commandments-of-clean-error-handling-in-typescript-93a9cbdf1af5#4424) to handle that (slim) possibility.

The type assertion (`as`) is convenient for stepping outside of TypeScript's bounds if you're confident you know what the type of a value will always be, but the drawback is you then lose the ability for TS to prevent some runtime errors in the case that assertion is incorrect.

File: src/config.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+const EMPTY_CONFIG: PluginConfig = {ruleOverrides: new Map()}
+
+export async function loadConfig(configPath: string, knownRuleIds: ReadonlySet<string>): Promise<PluginConfig> {
+  let raw: string
+  try {
+    raw = await readFile(configPath, 'utf8')
+  } catch (err) {
+    if ((err as NodeJS.ErrnoException)?.code !== 'ENOENT') {
+      console.error(`[alt-text-scan] failed to read ${configPath}; running with default rule settings:`, err)
```
Comment (JoyceZhu) [inline src/config.ts]: This wouldn't be an error, but a plain ol' `console.log()` (or `console.info()` if you want to be extra-precise). `console.error` goes into `stderr` IIRC, which is odd behavior when the developer didn't actually do anything wrong.

File: src/config.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    }
+    return EMPTY_CONFIG
+  }
+
+  let parsed: unknown
```
Comment (JoyceZhu) [inline src/config.ts:21]: I'd encourage using JSON Schema if possible to automatically validate the shape of the JSON, much easier and more reliable than manual validation and I think there's a good chance you'll see it in the future no matter where you work.

Here's an [example of a JSON Schema file in the big GitHub frontend repo](https://github.com/github/github-ui/blob/main/packages/ui-commands/schema/commands.schema.json), reliable [JSON Schema validator package](https://www.npmjs.com/package/json-schema-library), and if you're curious [their official overview of what it is](https://json-schema.org/overview/what-is-jsonschema).

Comment (kzhou314) [reply src/config.ts:21]: I looked a bit into it and noticed that if we were to automatically validate the JSON shape, it would require that the consumer install the validator when they load in the plugin (this was also why I chose to use a JSON rather than YAML file for the config, because likewise the consumer would have to install YAML on their end when loading in the plugin, since the plugins live outside the scanner's dependencies). This would add some friction for a plugin that is designed to just be dropped into the main scanner repo.

Alternatively, we could ship a schema/config.schema.json in this repo, and document "$schema" in the README so that consumers can still get the auto validation (autocomplete and inline errors)if they add an extra line to their config.

What are your thoughts on this?

Comment (JoyceZhu) [reply src/config.ts:21]: > Alternatively, we could ship a schema/config.schema.json in this repo, and document "$schema" in the README so that consumers can still get the auto validation (autocomplete and inline errors) if they add an extra line to their config.

Thanks for investigating further -- I would go with this approach, as it'll still be nice for users to have the _option_ to validate, and it's also nice to have the schema in a separate, readable file if we start adding more options.

**Comment by JoyceZhu** [comment]
@taarikashenafi, would you like to take a look through this PR as well?

File: index.ts [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
+// Config lives in the consumer's repo at `.github/scanner-plugins/<name>/config.json`.
+const configPath = join(process.cwd(), '.github', 'scanner-plugins', name, 'config.json')
+const knownRuleIds = new Set(allRules.map(r => r.id))
+const configPromise = loadConfig(configPath, knownRuleIds)
+
 export default async function altTextScan({page, addFinding}: PluginArgs): Promise<void> {
   const url = page.url()
 
+  const {ruleOverrides} = await configPromise
+  const enabledRules = allRules.filter(rule => ruleOverrides.get(rule.id) ?? rule.defaultEnabled ?? true)
```
Comment (taarikashenafi) [inline index.ts:19]: The filtering logic is correct, but we still need a test that proves it actually works. For example, disable a rule in config and assert addFinding is never called for it.

File: src/config.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
+import {readFile} from 'node:fs/promises'
+
+export type PluginConfig = {
+  ruleOverrides: ReadonlyMap<string, boolean>
+}
+
+const EMPTY_CONFIG: PluginConfig = {ruleOverrides: new Map()}
```
Comment (taarikashenafi) [inline src/config.ts]: EMPTY_CONFIG shares the same Map instance across all fallback returns. ReadonlyMap is TS-only so it's still mutable at runtime. Swap it for:


```suggestion
const emptyConfig = (): PluginConfig => ({ruleOverrides: new Map()})
```

File: src/config.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+export async function loadConfig(configPath: string, knownRuleIds: ReadonlySet<string>): Promise<PluginConfig> {
+  let raw: string
+  try {
+    raw = await readFile(configPath, 'utf8')
+  } catch (err) {
+    if (err instanceof Error && 'code' in err && err.code === 'ENOENT') {
+      return EMPTY_CONFIG
+    }
+    console.log(`[alt-text-scan] failed to read ${configPath}; running with default rule settings:`, err)
+    return EMPTY_CONFIG
```
Comment (taarikashenafi) [inline src/config.ts]: This should be `console.warn` instead of `console.log`. If a user has a permissions issue or some other read error, they need to actually see it, and `console.log` is too easy to miss compared to how the rest of the errors in this file are logged.

File: index.ts (hunk 4 above)
Comment (kzhou314) [reply index.ts:19]: Tested in the newest commit!

File: src/config.ts (hunk 6 above)
Comment (kzhou314) [reply src/config.ts]: That's a fair point, the console.log is inconsistent with how the other issues are logged. But Joyce mentioned in an earlier review that console.warn/error would go to `stderr` when the developer hasn't actually done anything wrong in this case. So the message technically wouldn't be in the right place. However, would be happy to revisit with her if you think this visibility issue outweighs the above concern.

Comment (JoyceZhu) [reply src/config.ts]: A possible differentiator is to consult the actual error code and direct the output to `stdout` or `stderr` (i.e. -- `log` vs `warn`) based on whether the error is from the file not existing (would be normal) or something else.

The `try / catch` block is what Node documentation recommends as the idiomatic way to deal with files, if you check out the [docs and examples for `fs.access`](https://nodejs.org/api/fs.html#fsaccesspath-mode-callback)

Comment (kzhou314) [reply src/config.ts]: Cool. Based on the error code now, I think this should be a console.warn then, since it is a non-ENOENT fail. I've updated this in the latest commit.

**Review by JoyceZhu — APPROVED** [review]
